offload_vio/device_tx/video_encoder.cpp

```diff
@@ -0,0 +1,189 @@
+
+    auto nvvideoconvert_0 = gst_element_factory_make("nvvideoconvert", "nvvideoconvert0");
+    auto nvvideoconvert_1 = gst_element_factory_make("nvvideoconvert", "nvvideoconvert1");
+
+    auto encoder_img0 = gst_element_factory_make("nvv4l2h264enc", "encoder_img0");
+    auto encoder_img1 = gst_element_factory_make("nvv4l2h264enc", "encoder_img1");
+
+    auto h265parse_0 = gst_element_factory_make("h265parse", "h265parse0");
+    auto h265parse_1 = gst_element_factory_make("h265parse", "h265parse1");
+
+    auto caps_8uc1 = gst_caps_new_simple("video/x-raw", "format", G_TYPE_STRING, "GRAY8", "framerate", GST_TYPE_FRACTION, 0, 1,
+                                         "width", G_TYPE_INT, IMG_WIDTH, "height", G_TYPE_INT, IMG_HEIGHT, NULL);
+    g_object_set(G_OBJECT(_appsrc_img0), "caps", caps_8uc1, nullptr);
+    g_object_set(G_OBJECT(_appsrc_img1), "caps", caps_8uc1, nullptr);
+    gst_caps_unref(caps_8uc1);
+
+    auto caps_convert_to = gst_caps_new_simple("video/x-raw", "format", G_TYPE_STRING, "NV12", "width", G_TYPE_INT, IMG_WIDTH,
+                                               "height", G_TYPE_INT, IMG_HEIGHT, NULL);
+    g_object_set(G_OBJECT(caps_filter_0), "caps", caps_convert_to, nullptr);
+    g_object_set(G_OBJECT(caps_filter_1), "caps", caps_convert_to, nullptr);
+    gst_caps_unref(caps_convert_to);
+
+    // set bitrate from environment variables
+    // g_object_set(G_OBJECT(encoder_img0), "bitrate", std::stoi(std::getenv("ILLIXR_BITRATE")), nullptr, 10);
+    // g_object_set(G_OBJECT(encoder_img1), "bitrate", std::stoi(std::getenv("ILLIXR_BITRATE")), nullptr, 10);
+    g_object_set(G_OBJECT(encoder_img0), "bitrate", ILLIXR_BITRATE, nullptr);
+    g_object_set(G_OBJECT(encoder_img1), "bitrate", ILLIXR_BITRATE, nullptr);
+
+    g_object_set(G_OBJECT(_appsrc_img0), "stream-type", 0, "format", GST_FORMAT_BYTES, "is-live", TRUE, nullptr);
+    g_object_set(G_OBJECT(_appsrc_img1), "stream-type", 0, "format", GST_FORMAT_BYTES, "is-live", TRUE, nullptr);
+
+    g_object_set(_appsink_img0, "emit-signals", TRUE, "sync", FALSE, nullptr);
+    g_object_set(_appsink_img1, "emit-signals", TRUE, "sync", FALSE, nullptr);
+
+    g_signal_connect(_appsink_img0, "new-sample", G_CALLBACK(cb_new_sample), this);
+    g_signal_connect(_appsink_img1, "new-sample", G_CALLBACK(cb_new_sample), this);
+
+    _pipeline_img0 = gst_pipeline_new("pipeline_img0");
+    _pipeline_img1 = gst_pipeline_new("pipeline_img1");
+
+    gst_bin_add_many(GST_BIN(_pipeline_img0), _appsrc_img0, nvvideoconvert_0, encoder_img0, h265parse_0, caps_filter_0,
+                     videoconvert_0, _appsink_img0, nullptr);
+    gst_bin_add_many(GST_BIN(_pipeline_img1), _appsrc_img1, nvvideoconvert_1, encoder_img1, h265parse_1, caps_filter_1,
+                     videoconvert_1, _appsink_img1, nullptr);
+
+    // link elements
+    if (!gst_element_link_many(_appsrc_img0, nvvideoconvert_0, encoder_img0, _appsink_img0, nullptr) ||
```

Comment:
When I wrote this part, I was experimenting with different components that weren't used in the final version. We should remove all of the components that are not linked together in the `gst_element_link_many` call. We should remove the definitions and the parameters in `gst_bin_add_many`.

Reply:
updated